gradle-plugin/src/proguard/gradle/configuration/ProGuardTaskConfiguration.java

```diff
@@ -0,0 +1,473 @@
+import org.gradle.api.tasks.PathSensitivity;
+import proguard.ClassPath;
+import proguard.ClassPathEntry;
+import proguard.ClassSpecification;
+import proguard.Configuration;
+import proguard.KeepClassSpecification;
+
+import java.io.File;
+import java.net.URL;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * A Gradle-aware wrapper for the ProGuard `Configuration`, allowing all inputs and outputs
+ * to be correctly declared. This class simply exposes the properties of `Configuration`, annotated
+ * correctly for the Gradle task with return types wrapped as appropriate.
+ *
+ * For clarity, fields on `Configuration` that are not task inputs or outputs are exposed with `@Internal`.
+ * These methods could just as easily be removed from this type.
+ */
+public class ProGuardTaskConfiguration {
+    protected final Provider<Configuration> configurationProvider;
+    private final ObjectFactory objectFactory;
+
+    public ProGuardTaskConfiguration(Provider<Configuration> configurationProvider, ObjectFactory objectFactory) {
+        this.configurationProvider = configurationProvider;
+        this.objectFactory = objectFactory;
+    }
+
+    public static ProGuardTaskConfiguration create(Provider<Configuration> configurationProvider, ObjectFactory objectFactory) {
+        if (GradleVersionCheck.isAtLeastGradle7()) {
+            return new ProGuardTaskConfiguration(configurationProvider, objectFactory);
+        } else {
+            return new ProGuardTaskConfigurationForGradle6(configurationProvider, objectFactory);
+        }
+    }
+
+    private Configuration getConfiguration() {
+        return configurationProvider.get();
+    }
+
+    protected <T> Provider<T> mappedProvider(Transformer<T, Configuration> transformer) {
+        return configurationProvider.map(transformer);
+    }
+
+    ///////////////////////////////////////////////////////////////////////////
+    // Input and output options.
+    ///////////////////////////////////////////////////////////////////////////
+
+    @Classpath
+    public FileCollection getInputJars() {
+        Provider<List<File>> inputJarsProvider = mappedProvider(config -> {
+            List<File> files = new ArrayList<>();
+            ClassPath programJars = config.programJars;
+            for (int i = 0; i < programJars.size(); i++) {
+                ClassPathEntry classPathEntry = programJars.get(i);
+                if (!classPathEntry.isOutput()) {
+                    files.add(classPathEntry.getFile());
+                }
+            }
+
+            ClassPath libraryJars = config.libraryJars;
+            for (int i = 0; i < libraryJars.size(); i++) {
+                ClassPathEntry classPathEntry = libraryJars.get(i);
+                if (!classPathEntry.isOutput()) {
+                    files.add(classPathEntry.getFile());
+                }
+            }
+            return files;
+        });
+        return objectFactory.fileCollection().from(inputJarsProvider);
```

Comment:
Is it important to create this thing on demand here? Can't you just create this file collection with the contents in the constructor?

Reply:
Fixed